block/manager.go

```diff
@@ -945,10 +956,34 @@ func (m *Manager) processVoteExtension(ctx context.Context, block *types.Block,
 	if err != nil {
 		return fmt.Errorf("error returned by ExtendVote: %w", err)
 	}
-	sign, err := m.proposerKey.Sign(extension)
-	if err != nil {
-		return fmt.Errorf("error signing vote extension: %w", err)
+
+	vote := &cmproto.Vote{
+		Height:    int64(newHeight),
+		Round:     0,
+		Extension: extension,
 	}
+	extSignBytes := cmtypes.VoteExtensionSignBytes(m.genesis.ChainID, vote)
+	var sign []byte
+	switch m.proposerKey.(type) {
```

Comment:
The entire switch statement is a huge code duplicate - please extract `sign` method.

Reply:
And it should be unit tested  - preferably sign + validate signature.